### PR TITLE
Fix Brotli http compression issue

### LINKED ISSUE #9969: Brotli HTTP compression failing on 20.3.5
We are getting the following exception when querying data using the Grafana ClickHouse plugin on the new 20.3.5 release. HTTP compression is enabled on the server, and the Grafana plugin sends an `Accept Encoding: gzip, deflate, br` header.  Is there some way we can diable brotli compression without disabling compression completely?

> 2020.03.31 13:54:50.167406 [ 26365 ] {15EF0662EB85DF85} <Error> executeQuery: Code: 449, e.displayText() = DB::Exception: brotli compress failed: While executing JSONRowOutputFormat (version 20.3.5.21 (official build)) (from [2001:558:fc0c:0:f816:3eff:fe6f:de33]:40234) (in que
> ry: SELECT     (intDiv(toUInt32(datetime), 15) * 15) * 1000 as t,     pipeline,     host,     avg(eps)    FROM analytics.kpipe_metrics_d WHERE datetime >= toDateTime(1585652088) and metric = 'processed' and pipeline like '%%%%' GROUP BY t, pipeline, host ORDER BY t FORMAT JSON
>  ), Stack trace (when copying this message, always include the lines below):
> 
> 0. Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x1050f0d0 in /usr/bin/clickhouse
> 1. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x8f3272d in /usr/bin/clickhouse
> 2. ? @ 0x8fd6a4c in /usr/bin/clickhouse
> 3. DB::WriteBufferFromHTTPServerResponse::nextImpl() @ 0x8fcd942 in /usr/bin/clickhouse
> 4. DB::WriteBufferValidUTF8::nextImpl() @ 0xaa7af5e in /usr/bin/clickhouse
> 5. DB::JSONRowOutputFormat::writeField(DB::IColumn const&, DB::IDataType const&, unsigned long) @ 0xdc81801 in /usr/bin/clickhouse
> 6. DB::IRowOutputFormat::write(std::__1::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn> > > const&, unsigned long) @ 0xdc0a4a9 in /usr/bin/clickhouse
> 7. DB::IRowOutputFormat::consume(DB::Chunk) @ 0xdc0a332 in /usr/bin/clickhouse
> 8. DB::IOutputFormat::work() @ 0xdbca30d in /usr/bin/clickhouse
> 9. ? @ 0xdbb2881 in /usr/bin/clickhouse
> 10. DB::PipelineExecutor::executeSingleThread(unsigned long, unsigned long) @ 0xdbb6b2d in /usr/bin/clickhouse
> 11. ? @ 0xdbb7522 in /usr/bin/clickhouse
> 12. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x8f567eb in /usr/bin/clickhouse
> 13. ? @ 0x8f54c33 in /usr/bin/clickhouse
> 14. start_thread @ 0x7dd5 in /usr/lib64/libpthread-2.17.so
> 15. __clone @ 0xfe02d in /usr/lib64/libc-2.17.so
>

### PR DESCRIPTION
**Pull request by kssenii**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Closes https://github.com/ClickHouse/ClickHouse/issues/9969. Fixed Brotli http compression error, which reproduced for large data sizes, slightly complicated structure and with json output format. Update Brotli to the latest version to include the "fix rare access to uninitialized data in ring-buffer".

Detailed description / Documentation draft:

Before this fix always got this error:

2021.02.19 20:05:05.200307 [ 2173111 ] {18f3278d-5fb4-4086-83d6-50228a0250f2} <Error> executeQuery: Code: 449, e.displayText() = DB::Exception: brotli compress failed: While exec
uting JSONRowOutputFormat (version 21.3.1.1) (from 127.0.0.1:45976) (in query: SELECT toDateTime(number + now()) as datetime, 'test-pipeline' as pipeline, 'clickhouse-test-host-0
01.clickhouse.com' as host, 'clickhouse' as home, 'clickhouse' as detail, 'clickhouse' as data FROM numbers(1000000) FORMAT JSON), Stack trace (when copying this message, always
include the lines below):

0. /home/kssenii/ClickHouse/build-asan/../contrib/libcxx/include/exception:133: Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::all
ocator<char> > const&, int) @ 0x25d179e9 in /home/kssenii/ClickHouse/build-asan/programs/clickhouse
1. /home/kssenii/ClickHouse/build-asan/../src/Common/Exception.cpp:56: DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, int, bool) @ 0xa8f9918 in /home/kssenii/ClickHouse/build-asan/programs/clickhouse
2. /home/kssenii/ClickHouse/build-asan/../src/IO/BrotliWriteBuffer.cpp:89: DB::BrotliWriteBuffer::nextImpl() @ 0x1c863cfa in /home/kssenii/ClickHouse/build-asan/programs/clickhouse

### CODE REVIEW

File: src/IO/BrotliWriteBuffer.cpp [hunk 1]
```diff
@@ -86,7 +86,7 @@ void BrotliWriteBuffer::nextImpl()
             throw Exception("brotli compress failed", ErrorCodes::BROTLI_WRITE_FAILED);
         }
     }
-    while (in_available > 0 || out_capacity == 0);
+    while (in_available > 0);
```
Comment (alexey-milovidov) [inline src/IO/BrotliWriteBuffer.cpp:89]: Do we have similar issue with zlib and zstd?

Btw, what is the meaning of this fix?

**Comment by alexey-milovidov** [comment]
> Fix #9969.

Let's add more detailed changelog entry.

**Comment by alexey-milovidov** [comment]
https://clickhouse-test-reports.s3.yandex.net/20991/0d88366b2775bdcb60ae3eb18bc9fcb2ce7eef01/functional_stateless_tests_(memory)/stderr.log

**Comment by kssenii** [comment]
> https://clickhouse-test-reports.s3.yandex.net/20991/0d88366b2775bdcb60ae3eb18bc9fcb2ce7eef01/functional_stateless_tests_(memory)/stderr.log

For now I figured out that this msan failure can be fixed by changing compression level, but about 3% of last rows get lost, but without changing compression level it did not encode even half before failed with msan report.  And msan failures are not connected with the fix for the issue of this pr, because without this fix, it also fails with msan error without having encoded even half of data.

**Comment by alexey-milovidov** [comment]
Need to check if Brotli is using uninitialized memory legitimately.

**Comment by kssenii** [comment]
What if with the given arguments passed to method, where uninitialised data access is done, it is impossible to correctly define up to which position array is initialized (to add a check)? (My attempts fix msan, but some rows are lost). And Brotli always initializes less than it allocates for that array, but does it with non-obvious calculations.

**Comment by alexey-milovidov** [comment]
There are at least two possibilities:
- we pass uninitialized data to brotli or using its interface incorrectly;
- brotli is using uninitizlied memory in legitimate way in its algorithms and it is not annotated for memory sanitizer (unlikely).

**Comment by kssenii** [comment]
> https://clickhouse-test-reports.s3.yandex.net/20991/0d88366b2775bdcb60ae3eb18bc9fcb2ce7eef01/functional_stateless_tests_(memory)/stderr.log

hooray, `Functional stateless tests (memory)` passed.

> Do we have similar issue with zlib and zstd?

`zlib` and `zstd` do not have the same while condition for compression and the same long test with json format works fine for them (checked with asan and msan). Should I add this test for them in this pr too?

> Btw, what is the meaning of this fix?

Before this fix with a while condition `while (in_available > 0 || out_capacity == 0)` Brotli always failed if after `in_available == 0 && out_capacity == 0` it went into another compression iteration. in_available is decremented by the amount of bytes consumed and out_capacity is decremented by the amount of bytes written, so if they both are 0 after an iteration, we should not continue to compress in a loop, because there is nothing left.

**Comment by alexey-milovidov** [comment]
Ok.

> The same long test with json format works fine for zlib and zstd (checked with asan and msan). Should I add this test for them in this pr too?

Not necessarily, it's enough.

**Comment by alexey-milovidov** [comment]
@kssenii Maybe add similar tests for zstd and zlib just in case, in separate PR?

**Comment by alexey-milovidov** [comment]
> AST fuzzer (debug) — Logical error: 'Logical error: no information about file statusMap%2Egoal_id.size0 in StorageLog'.

An issue with nested columns, @CurtizJ.

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**